### PR TITLE
Refactor models and controllers for better readability

### PR DESCRIPTION
**Pull request by Bodacious**
This PR contains a selection of refactorings in the models, helpers, views and controllers.

Most of these changes came from the spec-writing work, but have been committed on a separate branch to prevent false positive tests where tests are passing because the code has changed.

I'll add notes inline on this PR to explain each change

### CODE REVIEW

File: app/controllers/answers_controller.rb [hunk 1]
```diff
@@ -56,7 +56,7 @@ def create_or_update
         }
       }).find(p_params[:plan_id])
       @question = @answer.question
-      @section = @plan.get_section(@question.section_id)
+      @section = @plan.sections.find_by(id: @question.section_id)
```
Comment (Bodacious) [inline app/controllers/answers_controller.rb]: This method was unnecessary. Refactored to use built-in Rails association.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -96,64 +96,53 @@ def success_message(obj_name, action)
     "#{_('Successfully %{action} your %{object}.') % {object: obj_name, action: action}}"
   end
 
-  # Check whether the string is a valid array of JSON objects
-  def is_json_array_of_objects?(string)
```
Comment (Bodacious) [inline app/controllers/application_controller.rb]: As far as I can see, `is_json_array_of_objects?` is not used within the code. Removed.

File: app/controllers/concerns/conditional_user_mailer.rb [hunk 3]
```diff
@@ -1,35 +1,27 @@
+# frozen_string_literal: true
+
 module ConditionalUserMailer
-  extend ActiveSupport::Concern
```
Comment (Bodacious) [inline app/controllers/concerns/conditional_user_mailer.rb]: `ActiveSupport::Concern` was not being used correctly here. The included block is a callback, and is not necessary to include methods within another class (that's the default behaviour of Modules in Ruby).

Please read the documentation for [ActiveSupport::Concern](https://api.rubyonrails.org/v5.2/classes/ActiveSupport/Concern.html)

File: app/controllers/concerns/conditional_user_mailer.rb [hunk 4]
```diff
@@ -1,35 +1,27 @@
...
 
-  # Adds following methods as class methods for the class that include this module
-  included do
```
Comment (Bodacious) [inline app/controllers/concerns/conditional_user_mailer.rb]: `ActiveSupport::Concern` was not being used correctly here. The included block is a callback, and is not necessary to include methods within another class (that's the default behaviour of Modules in Ruby).

Please read the documentation for [ActiveSupport::Concern](https://api.rubyonrails.org/v5.2/classes/ActiveSupport/Concern.html) for tips on correct usage.

File: app/controllers/concerns/versionable.rb [hunk 5]
```diff
@@ -1,106 +1,123 @@
 module Versionable
-  extend ActiveSupport::Concern
 
-  included do
```
Comment (Bodacious) [inline app/controllers/concerns/versionable.rb]: `ActiveSupport::Concern` was not being used correctly here. The included block is a callback, and is not necessary to include methods within another class (that's the default behaviour of Modules in Ruby).

Please read the documentation for [ActiveSupport::Concern](https://api.rubyonrails.org/v5.2/classes/ActiveSupport/Concern.html) for tips on correct usage.

File: app/controllers/org_admin/plans_controller.rb [hunk 6]
```diff
@@ -40,20 +40,20 @@ def download_plans
...
-            "#{plan.title}", 
-            "#{plan.template.title}", 
-            "#{plan.owner.org.present? ? plan.owner.org.name : ''}", 
+            "#{plan.title}",
+            "#{plan.template.title}",
+            "#{plan.owner.org.present? ? plan.owner.org.name : ''}",
             "#{plan.owner.name(false)}",
             "#{plan.owner.email}",
             "#{l(plan.latest_update.to_date, formats: :short)}",
-            "#{Plan.visibility_message(plan.visibility.to_sym).capitalize}"
```
Comment (Bodacious) [inline app/controllers/org_admin/plans_controller.rb]: Replaced this method with a constant for better performance.

File: app/controllers/plans_controller.rb [hunk 7]
```diff
@@ -168,7 +168,6 @@ def edit
     render('/phases/edit', locals: {
       base_template_org: phase.template.base_org,
       plan: plan, phase: phase, readonly: readonly,
-      question_guidance: plan.guidance_by_question_as_hash,
```
Comment (Bodacious) [inline app/controllers/plans_controller.rb]: Removed this method. Doesn't seem to be used in the code, and the variable set here doesn't seem to be used in the views.

File: app/controllers/plans_controller.rb [hunk 8]
```diff
@@ -237,9 +236,6 @@ def destroy
   def status
     @plan = Plan.find(params[:id])
     authorize @plan
-    respond_to do |format|
-      format.json { render json: @plan.status }
```
Comment (Bodacious) [inline app/controllers/plans_controller.rb]: Removed this block. Endpoint doesn't seem to be used at all.

File: app/controllers/plans_controller.rb [hunk 9]
```diff
@@ -405,25 +401,6 @@ def get_most_recent( templates )
     groups.values
   end
 
-
-  def fixup_hash(plan)
```
Comment (Bodacious) [inline app/controllers/plans_controller.rb]: Remove this unused method.

File: app/controllers/roles_controller.rb [hunk 10]
```diff
@@ -9,7 +9,7 @@ def create
     authorize @role
 
     access_level = params[:role][:access_level].to_i
-    @role.set_access_level(access_level)
```
Comment (Bodacious) [inline app/controllers/roles_controller.rb]: Replaced this method with a ruby idiomatic setter.

File: app/controllers/roles_controller.rb [hunk 11]
```diff
@@ -51,7 +51,7 @@ def update
     @role = Role.find(params[:id])
     authorize @role
     access_level = params[:role][:access_level].to_i
-    @role.set_access_level(access_level)
```
Comment (Bodacious) [inline app/controllers/roles_controller.rb]: Replaced this method with a ruby idiomatic setter.

File: app/helpers/application_helper.rb [hunk 12]
```diff
@@ -13,11 +13,6 @@ def resource
   def devise_mapping
     @devise_mapping ||= Devise.mappings[:user]
   end
-  
-  # ---------------------------------------------------------------------------
-  def hash_to_js_json_variable(obj_name, hash)
```
Comment (Bodacious) [inline app/helpers/application_helper.rb]: Removed this unused method

File: app/helpers/application_helper.rb [hunk 13]
```diff
@@ -30,10 +25,6 @@ def isActivePage(path, exact_match = false)
     end
   end
 
-  def is_integer?(string)
```
Comment (Bodacious) [inline app/helpers/application_helper.rb]: Removed this unused method.

File: app/helpers/plans_helper.rb [hunk 14]
```diff
@@ -1,27 +1,10 @@
 module PlansHelper
-  # Shows whether the user has default, template-default or custom settings
-  # for the given plan.
-  # --------------------------------------------------------
-  def plan_settings_indicator(plan)
```
Comment (Bodacious) [inline app/helpers/plans_helper.rb]: Removed this unused method.

**Comment by Bodacious** [comment]
Please ignore this PR for now, until #1756 has been concluded.

File: app/views/user_mailer/plan_visibility.html.erb [hunk 15]
```diff
@@ -2,7 +2,7 @@
   tool_name = Rails.configuration.branding[:application][:name]
   username = @user.name
   plan_title = @plan.title
-  plan_visibility = Plan.visibility_message(@plan.visibility.to_sym)
```
Comment (Bodacious) [inline app/views/user_mailer/plan_visibility.html.erb]: Replaced this with a constant for better performance.

File: app/views/user_mailer/permissions_change_notification.html.erb [hunk 16]
```diff
@@ -4,14 +4,18 @@
...
   access = nil
   access_level = @role.access_level()
-  access_level_messages = Role.access_level_messages
 %>
 <% FastGettext.with_locale FastGettext.default_locale do %>
   <p>
     <%= _('Hello %{username}') %{ :username => username } %>
   </p>
   <p>
-    <%= _('Your permissions relating to %{plan_title} have changed. You now have %{type} access. This means you can %{placeholder1} %{placeholder2}') %{ :plan_title => plan_title, :type => access_level_messages[access_level][:type], :placeholder1 => access_level_messages[access_level][:placeholder1], :placeholder2 => access_level_messages[access_level][:placeholder2] } %>
```
Comment (Bodacious) [inline app/views/user_mailer/permissions_change_notification.html.erb]: Replaced with constant for better performance

File: app/views/phases/edit.html.erb [hunk 17]
```diff
@@ -1,8 +1,7 @@
-<%# locals: { plan, phase, readonly, question_guidance } %>
```
Comment (Bodacious) [inline app/views/phases/edit.html.erb]: Remove unused comment

File: app/views/phases/_edit_plan_answers.html.erb [hunk 18]
```diff
@@ -1,4 +1,3 @@
-<%# locals: { plan, phase, readonly, question_guidance, guidance_groups } %>
```
Comment (Bodacious) [inline app/views/phases/_edit_plan_answers.html.erb]: Removed unused comment.

File: app/views/org_admin/phases/preview.html.erb [hunk 19]
```diff
@@ -36,7 +36,6 @@
                     plan: nil,
                     phase: phase,
                     readonly: true,
-                    question_guidance: {},
```
Comment (Bodacious) [inline app/views/org_admin/phases/preview.html.erb]: Removed unused variable.

File: app/models/user.rb [hunk 20]
```diff
@@ -172,14 +184,6 @@ def identifier_for(scheme)
     user_identifiers.where(identifier_scheme: scheme).first
   end
 
-  ##
-  # sets a new organisation for the user
-  #
-  # @param new_organisation [Organisation] the new organisation for the user
-  def organisation=(new_org)
```
Comment (Bodacious) [inline app/models/user.rb]: Removed unused method.

File: app/models/user.rb [hunk 21]
```diff
@@ -154,15 +175,6 @@ def name(use_email = true)
     end
   end
 
-  ##
-  # returns all active plans for a user
-  #
-  # @return [Plans]
-  def active_plans
```
Comment (Bodacious) [inline app/models/user.rb]: Removed unused method.

File: app/models/user.rb [hunk 22]
```diff
@@ -63,39 +66,49 @@ class User < ActiveRecord::Base
...
   has_one  :pref
+
   has_many :answers
+
   has_many :notes
+
   has_many :exported_plans
+
   has_many :roles, dependent: :destroy
-  has_many :plans, through: :roles do
```
Comment (Bodacious) [inline app/models/user.rb]: Removed unused metod in association

File: app/models/template.rb [hunk 23]
```diff
@@ -61,53 +100,28 @@ class Template < ActiveRecord::Base
...
       unarchived.where(published: true)
     end
   }
 
-  # Retrieves the latest templates, i.e. those with maximum version associated.
-  # It can be filtered down if family_id is passed. NOTE, the template objects
-  # instantiated only contain version and family attributes populated. See
-  # Template::latest_version scope method for an adequate instantiation of
-  # template instances.
-  scope :latest_version_per_family, -> (family_id = nil) {
```
Comment (Bodacious) [inline app/models/template.rb]: Moved this scope to a private method.

File: app/models/template.rb [hunk 24]
```diff
@@ -121,24 +135,24 @@ class Template < ActiveRecord::Base
 
   # Retrieves the latest templates, i.e. those with maximum version associated
   # for a set of org_id passed
-  scope :latest_version_per_org, -> (org_id = nil) {
-    if org_id.respond_to?(:each)
```
Comment (Bodacious) [inline app/models/template.rb]: Refactor how local variable is defined with conditional

File: app/models/template.rb [hunk 25]
```diff
@@ -61,53 +100,28 @@ class Template < ActiveRecord::Base
 
   scope :unarchived, -> { where(archived: false) }
 
-  scope :default, -> { where(is_default: true, published: true).last }
```
Comment (Bodacious) [inline app/models/template.rb]: Moved this to class method. It's not a scope. Scopes should be used to filter records. This method returns a `Template`, not an `ActiveRecord::Relation`

File: app/models/template.rb [hunk 26]
```diff
@@ -148,126 +162,153 @@ class Template < ActiveRecord::Base
...
   # defines the export setting for a template object
   has_settings :export, class_name: 'Settings::Template' do |s|
     s.key :export, defaults: Settings::Template::DEFAULT_SETTINGS
   end
 
-  validates :org, :title, presence: {message: _("can't be blank")}
+  validates :org, :title, presence: { message: _("can't be blank") }
 
-  # Class methods gets defined within this
-  class << self
```
Comment (Bodacious) [inline app/models/template.rb]: Redefined class methods using more explicit syntax (e.g. `def self.my_method`)

File: app/controllers/org_admin/phases_controller.rb [hunk 27]
```diff
@@ -127,7 +127,7 @@ def update
     def sort
       @phase = Phase.find(params[:id])
       authorize @phase
-      Section.update_numbers!(*params.fetch(:sort_order, []), phase: @phase)
```
Comment (Bodacious) [inline app/controllers/org_admin/phases_controller.rb]: Rename this variable to make sortable general.

File: app/models/concerns/acts_as_sortable.rb [hunk 28]
```diff
@@ -0,0 +1,44 @@
+module ActsAsSortable
```
Comment (Bodacious) [inline app/models/concerns/acts_as_sortable.rb]: Added this module to share the `update_numbers!` method across multiple models

File: app/models/concerns/exportable_plan.rb [hunk 29]
```diff
@@ -1,122 +1,120 @@
...
-        end
-        exported_plan.save
+      Settings::Template::DEFAULT_SETTINGS.each do |key, value|
+        ep.settings(:export).send("#{key}=", plan_settings.send(key))
       end
+    end
+    exported_plan.save
+  end
 
-      def sanitize_text(text)
```
Comment (Bodacious) [inline app/models/concerns/exportable_plan.rb]: Replaced nil check with `to_s`. **Tell, don't ask!**

File: app/models/concerns/json_link_validator.rb [hunk 30]
```diff
@@ -1,32 +1,15 @@
 module JSONLinkValidator
-  extend ActiveSupport::Concern
 
-  included do
-    # Parses a stringified JSON according to validate_links (e.g. [{ link: String, text: String}, ...])
-    # param {String} the stringified JSON value
-    # Returns an Array of hashes after decoding/validating the stringified JSON passed, otherwise nil
-    def parse_links(value)
```
Comment (Bodacious) [inline app/models/concerns/json_link_validator.rb]: Remove this unused method.

File: app/models/question_format.rb [hunk 31]
```diff
@@ -12,34 +12,49 @@
...
-  validates :title, presence: {message: _("can't be blank")}, uniqueness: {message: _("must be unique")}
+  # ===============
+  # = Validations =
+  # ===============
 
+  validates :title, presence: { message: PRESENCE_MESSAGE },
+                    uniqueness: { message: UNIQUENESS_MESSAGE }
 
-  # Retrieves the id for a given formattype passed
-  scope :id_for, -> (formattype) { where(formattype: formattype).pluck(:id).first }
```
Comment (Bodacious) [inline app/models/question_format.rb]: This is not a scope. Refactored as a class method.

File: app/models/role.rb [hunk 32]
```diff
@@ -80,52 +143,17 @@ def access_level
     end
   end
 
-  # Sets access_level according to bit fields defined in the column access
-  # TODO refactor according to the hash defined above (e.g. 1 key is :creator, 2 key is :administrator, etc)
-  def set_access_level(access_level)
```
Comment (Bodacious) [inline app/models/role.rb]: Deprecated this method and replaced with an idiomatic Ruby setter method.

File: app/controllers/plans_controller.rb (hunk 8 above)
Comment (xsrust) [reply app/controllers/plans_controller.rb]: If the endpoint isnt removed, could we remove the route/action alltogether?

File: app/models/concerns/acts_as_sortable.rb (hunk 28 above)
Comment (xsrust) [reply app/models/concerns/acts_as_sortable.rb]: Thanks, it's good to get behavior like this generic!

File: app/models/role.rb (hunk 32 above)
Comment (xsrust) [reply app/models/role.rb]: that's much cleaner

File: app/controllers/plans_controller.rb (hunk 7 above)
Comment (briri) [reply app/controllers/plans_controller.rb]: thanks for catching that. we must have missed it in the rewrite of the pages

File: app/controllers/plans_controller.rb (hunk 8 above)
Comment (briri) [reply app/controllers/plans_controller.rb]: Hmm. I think this action can probably be removed. There used to be a separate ajax call to update the plan status but I think it was refactored and became part of the json response returned back by the update/create actions of the answers_controller

**Comment by Bodacious** [comment]
Closes #1572

**Pull request merged**